File: cups/src/binding.cpp
// Libcups v2 docs:
// https://openprinting.github.io/cups/doc/cupspm.html
// Libcups v3 docs:
// https://openprinting.github.io/cups/libcups/cupspm.html
// node-addon-api docs:
// https://github.com/nodejs/node-addon-api/tree/main/doc

#include <napi.h>

#include "create-connection.h"
#include "dest-check-mime-type-support.h"
#include "dest-get-info.h"
#include "dest-get-job-creation-attributes.h"
#include "dest-send-job.h"
#include "get-dests.h"

Napi::Object init(Napi::Env env, Napi::Object exports) {
  exports.Set("createConnection", Napi::Function::New(env, createConnection));
  exports.Set("getDests", Napi::Function::New(env, getDests));
  exports.Set("destGetInfo", Napi::Function::New(env, destGetInfo));
  exports.Set("destCheckMimeTypeSupport",
              Napi::Function::New(env, destCheckMimeTypeSupport));
  exports.Set("destGetJobCreationAttributes",
              Napi::Function::New(env, destGetJobCreationAttributes));
  exports.Set("destSendJob", Napi::Function::New(env, destSendJob));

  return exports;
}

NODE_API_MODULE(cups, init)

File: cups/src/ts/binding.ts
import { createRequire } from "node:module";
import type { CupsConnectionData, CupsPrinterData, CupsPrinterInfoData } from "./index.ts";
import type { JobCreationAttributesAvailable, JobCreationAttributesSelected, JobId } from "./utils.ts";

const require = createRequire(import.meta.url);
const lib = require("../../build/cups.node");

export function createConnection(url: string): Promise<CupsConnectionData> {
  return lib.createConnection(url);
}

export function getDests(connection: CupsConnectionData): Promise<{ name: string; data: CupsPrinterData }[]> {
  return lib.getDests(connection);
}

export function destGetInfo(connection: CupsConnectionData, dest: CupsPrinterData): Promise<CupsPrinterInfoData> {
  return lib.destGetInfo(connection, dest);
}

export function destGetJobCreationAttributes(
  connection: CupsConnectionData,
  dest: CupsPrinterData,
  info: CupsPrinterInfoData,
): JobCreationAttributesAvailable {
  const attributes: JobCreationAttributesAvailable = lib.destGetJobCreationAttributes(connection, dest, info);
  // Modern CUPS proxies can automatically handle any number of copies.
  attributes.push({ name: "copies", type: "number", constraints: { min: 1, max: 9999 }, default: 1 });
  return attributes;
}

export async function destSendJob(
  connection: CupsConnectionData,
  dest: CupsPrinterData,
  info: CupsPrinterInfoData,
  title: string,
  options: JobCreationAttributesSelected,
  documentMimeType: string,
  documentBuffer: Uint8Array,
): Promise<JobId> {
  return lib.destSendJob(connection, dest, info, title, options, documentMimeType, documentBuffer);
}
